### PR TITLE
average odds

### PR DESCRIPTION
**Pull request by kschweig**
Adding average odds as a new fairness constraint

### CODE REVIEW

File: error_parity/cvxpy_utils.py [hunk 1]
```diff
@@ -375,6 +376,25 @@ def group_positive_prediction_rate(group_idx: int):
...
+        def average_tpr_fpr(group_idx: int):
+            group_tpr = groupwise_roc_points_vars[group_idx][1]
+            group_fpr = groupwise_roc_points_vars[group_idx][0]
+
+            return (group_tpr + group_fpr) / 2
+
+        # Add constraints on the absolute difference between group-wos
+        constraints += [
+            cp.abs(
+                average_tpr_fpr(i) - average_tpr_fpr(j)
```
Comment (AndreFCruz) [inline error_parity/cvxpy_utils.py:392]: If I'm interpreting this correctly, then this metric definition is based on equality of $(TPR+FPR)/2$ between all groups; are you sure @kschweig?

I thought it would be the same as equalized odds but using average to aggregate differences instead of max.

**Comment by AndreFCruz** [comment]
Thanks for the PR!

I left a single question on the metric's definition.
Do you have a reference for a paper that defines the equalized odds violation this way?

I saw that the FairShap paper uses a similar definition but they seem to define it for binary groups; I'm wondering if for multiple groups we'd just ensure it with max-avg:
$\max_{a,b\in S} \left[ \text{avg} (|FPR_a-FPR_b|, |TPR_a - TPR_b|) \right]$

File: error_parity/cvxpy_utils.py (hunk 1 above)
Comment (kschweig) [reply error_parity/cvxpy_utils.py:392]: This is indeed a **relaxed version** of equalized odds, where instead of the two constraints $P(\hat{Y} \mid A=1, Y=y) = P(\hat{Y} \mid A=0, Y=y) \quad y \in \\{0, 1\\}$ a single constraint $E_{Y}[P(\hat{Y} \mid A=1, Y)] = E_{Y}[P(\hat{Y} \mid A=0, Y)]$ is defining it. The empirical estimator for this constraint is $(\Delta TPR + \Delta FPR) / 2$. This is especially inspired by AIF360 offering the average odds difference as a metric: https://aif360.readthedocs.io/en/stable/modules/generated/aif360.metrics.ClassificationMetric.html#aif360.metrics.ClassificationMetric.average_odds_difference however I have no link to a paper investigating this metric in depth. Loosely speaking it allows to discriminate one group in TPR if the other is equally discriminated in FPR.

Comment (AndreFCruz) [reply error_parity/cvxpy_utils.py:392]: thanks for the clarification!

It's a bit of a weird metric because if one group has lower TPR and higher FPR than it can seen as zero constraint violation `average_odds=0.0`; but that would just be the worse of both metrics.

So in fact, a value of `0.0` constraint violation is not strictly equalized odds as there are infinite other solutions to this constraint by balancing higher FPR with lower TPR, or lower FPR with higher TPR.

**Comment by AdrianArnaiz** [comment]
> Thanks for the PR!
> 
> I left a single question on the metric's definition. Do you have a reference for a paper that defines the equalized odds violation this way?
> 
> I saw that the [FairShap](https://arxiv.org/pdf/2303.01928.pdf) paper uses a similar definition (above Eq. 5) but they seem to define it for binary groups; I'm wondering if for multiple groups we'd just ensure it with max-avg: $max_{a,b∈S}[avg(|FPRa−FPRb|,|TPRa−TPRb|)]$

*AOD* - Very interesting discussion and I agree that and AOD=0 does not necessarily fulfills the strict definition of Eq Odds $P(\hat{Y}=1 \mid A=1, Y=y) = P(\hat{Y}=1 \mid A=0, Y=y) \quad y \in \\{0, 1\\}$. I see the application of the AOD metric in a real-world scenario where the cost of the TPR failures could be compensated with the FPR failures in the other direction. While it might suggest equal treatment in terms of average discrimination, it doesn't necessarily mean equal treatment in all scenarios. For instance, in some cases, it might be more acceptable to have slightly higher false positives if it significantly reduces false negatives, and vice versa.

*References* - We (with @kschweig) found usage of AOD in different papers [[1](https://arxiv.org/pdf/2107.08310.pdf), [2](https://dl.acm.org/doi/pdf/10.1145/3468264.3468537), [3](https://www.jair.org/index.php/jair/article/view/13197), etc] and pages [[a](https://dataplatform.cloud.ibm.com/docs/content/wsj/model/wos-fairness-metrics-ovr.html?context=cpdaas),[b](https://www.mathworks.com/help/risk/explore-fairness-metrics-for-credit-scoring-model.html)], but they all refer to the **[AIF360 companion paper](https://arxiv.org/pdf/1810.01943.pdf)**.

*Multigroup* - I think your extension to multigroup is a valid approach. You either could use your proposed min-max approach or the *general* approach $\max_{a,b∈S}[\text{avg}(|FPR_a−FPR|,|TPR_a−TPR|)]$ where FPR and TPR are computed using all the datapoints.

**Comment by AndreFCruz** [comment]
Right; I agree that average-odds could make sense if it used average FPR and FNR differences, i.e.:

$\max_{a,b\in S} \left[ \text{avg} (FPR_a-FPR_b, FNR_a - FNR_b) \right]$

But since the aif360 definition uses FPR differences and TPR differences, then really it allows you to discriminate on both sides (lower TPR and higher FPR is deemed fair); for example, in the following plot, given the ROC point of group A, group B can lie on any point on the line between A and the diagonal and it would still be considered fair: 
IMG_0242

Just wanted to make sure I got this right because the aif360 package seems to be wrong by listing it as a relaxation of equalized-odds, since `average_odds=0` does not imply `equalized_odds=0`.

I think I'll just change the code slightly to use average of absolute differences as that is indeed a relaxation of equalized-odds; and that was the one used on the FairSHAP paper and a few others.

Thanks for the input and code contribution!

**Comment by kschweig** [comment]
Indeed, I fully agree that absolute differences make more sense, as it circumvents the issue of balancing higher FPR and lower TPR within the group. Thank you for taking so much time for investigating this PR!

As a side note, AIF360 has an absolute version as well (https://aif360.readthedocs.io/en/stable/modules/generated/aif360.sklearn.metrics.average_odds_error.html) with identical description, which is therefore correct.

Given the wide adoption of AIF360, we should consider opening an issue to raise awareness to this discrepancy.

**Comment by AndreFCruz** [comment]
Right, thanks for the sanity check confirmation!

Then it seems it's just IBM that's confusing people when stating that "At 0: Both groups have equal odds." for the non-absolute-value average-odds metric 😅 

I'll open an issue with AIF360 👍
